build/configure/app_configure_step.js

```diff
@@ -0,0 +1,149 @@
+   *                                   building steps.
+   */
+  start: function() {
+    var options = this.options;
+    this._preSteps = [];
+    this.appDirFile = utils.getFile(this.appDir);
+    this.xpcshell = utils.getEnv('XPCSHELLSDK');
+
+    // we'll execute our current modules with stringified options.
+    this.appOptions = utils.cloneJSON(options);
+
+    this.appOptions.APP_DIR = this.appDir,
+    this.appOptions.STAGE_APP_DIR = utils.joinPath(this.options.STAGE_DIR,
+      this.appDirFile.leafName);
+    this.appOptions.webapp = this.webapp;
+    // We may encouter error while stringify manifest , put it to command line
+    // and parse it back to json object, so we simply delete these since the
+    // following steps don't need these attributes.
+    delete this.appOptions.webapp.manifest;
+    delete this.appOptions.webapp.metaData;
+
+    utils.ensureFolderExists(utils.getFile(this.appOptions.STAGE_APP_DIR));
+    // Create app's own makefile and store in build_stage/mks.
+
+    this.buildConfig = new BuildConfig(
+      utils.joinPath(this.appOptions.STAGE_APP_DIR, 'build_config.in'),
+      utils.joinPath(options.GAIA_DIR, 'build_config.in')
+    );
+
+    this.buildConfig.insertConfig('APP_OPTIONS',
+      JSON.stringify(this.appOptions).replace(/\\/g, '\\\\')
+      .replace(/"/g, '\\"'));
+
+    // We have to rewrite below two ENV since they're using abspath and CURDIR.
+    this.buildConfig.insertConfig('GAIA_DIR',
+      options.GAIA_DIR);
+    this.buildConfig.insertConfig('XPCSHELLSDK',
+      options.XPCSHELLSDK);
+
+    this.appMkPath = utils.joinPath(this.appOptions.STAGE_APP_DIR, 'app.mk');
+    this.mainMake = new DependencyGraph(this.appMkPath);
+
+    this.genAppConfig();
+    this.buildConfig.writeConfig();
+
+    this.mainMake.genBackend('-include ' + this.buildConfig.path + '\n');
+  },
+
+  get execMakeCommand() {
+    return '@make -f ' + this.appMkPath;
+  },
+
+  /**
+   * In order to run buildscript under each app, we need to use below command.
+   * XXX: we should try to fix it in xpcshell-commonjs.js.
+   *
+   * @param {string} moduleName - the module we would like to execute.
+   */
+  templateCommand: function(moduleName) {
+    return '@$(XPCSHELLSDK) -f $(GAIA_DIR)/build/xpcshell-commonjs.js ' +
+    '-e "run(\'' + moduleName + '\',$(APP_OPTIONS));"';
+  },
+
+  // Get the last building steps of the app.
+  get lastSteps() {
+    return this._preSteps;
+  },
+
+  /**
+   * Generate app's own config data for buildbackend.
+   */
+  genAppConfig: function() {
+    var buildFile = utils.getFile(this.appDir, 'build', 'build.js');
+    if (buildFile.exists()) {
+      // execute app's build.js if it has.
+      this.mainMake.insertTask(
+        'phony',
+        this.webapp.sourceDirectoryName + '_build',
+        this._preSteps,
+        '@echo building ' + this.webapp.sourceDirectoryName + '\n\t' +
+        this.templateCommand('build-app')
```

Comment:
I think we should invoke <app>/build/build.js directly instead of build-app. Because build-app will trigger post-app which means after post-app is triggered build-app and then app.mk will trigger it again.